### PR TITLE
[SPARK-35648][PYTHON] Refine and add dependencies needed for dev in dev/requirement.txt

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR proposes to update `dev/requirement.txt` file.

NOTE that:
- This file isn't used anywhere in Apache Spark CI. It's just for convenience
- To minimize the overhead of maintenance, I removed all lowerbounds of dependencies, which means that using the latest versions of them should work in the clean environment (e.g., you can reinstall all of them).

### Why are the changes needed?

To note the dependencies needed for Spark dev, and for easier env setting up.

### Does this PR introduce _any_ user-facing change?

No, dev-only.

### How was this patch tested?

Logically derived from setup.py, and other places like CI

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @srowen @dongjoon-hyun FYI

**Review by dongjoon-hyun — APPROVED** [review]
+1. I love this. :)

**Comment by HyukjinKwon** [comment]
I added some more comments!

**Review by dongjoon-hyun — APPROVED** [review]

File: dev/requirements.txt [hunk 1]
```diff
@@ -1,8 +1,34 @@
...
+# Documentation (Python)
 pydata_sphinx_theme
 ipython
 nbsphinx
 numpydoc
+jinja2<3.0.0
+sphinx<3.1.0
+sphinx-plotly-directive
+
+# Developement scripts
```
Comment (dongjoon-hyun) [inline dev/requirements.txt:32]: nit. Typo?

Comment (HyukjinKwon) [inline dev/requirements.txt:32]:
```suggestion
# Development scripts
```

Comment (HyukjinKwon) [reply dev/requirements.txt:32]: oops

**Comment by HyukjinKwon** [comment]
Merging it in. It's not sued in CI.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Review by srowen — COMMENTED** [review]
Lgtm

**Pull request closed without merging**